### ISSUE TITLE
azurerm_virtual_machine storage_data_disk create_option = "Attach" cannot be resized

### ISSUE THREAD

**Issue by jake-delorme** [terraform-providers/terraform-provider-azurerm#345]
### Terraform Version

Terraform v0.10.5

### Affected Resource(s)
- storage_data_disk
- azurerm_managed_disk

### Terraform Configuration Files
```hcl
resource "azurerm_managed_disk" "data_disk" {
  count = "${var.servers}"
  name = "data1"
  resource_group_name = "${var.resource_group}"
  location = "${var.location}"
  storage_account_type = "${var.disk_type}"
  create_option = "Empty"
  disk_size_gb = "${var.data_disk_size}"
}

resource "azurerm_virtual_machine" "vm" {
  count = "${var.servers}"
  ...
  storage_data_disk {
    name              = "${element(azurerm_managed_disk.data_disk.*.name, count.index)}"
    managed_disk_id   = "${element(azurerm_managed_disk.data_disk.*.id, count.index)}"
    disk_size_gb = "${element(azurerm_managed_disk.data_disk.*.disk_size_gb, count.index)}"
    create_option     = "Attach"
    lun               = 0
  }
  ...
}
```

### Expected Behavior
When increasing the size of the data disk its expected that the attached disk would increase in size and be seen by the VM its attached to.

### Actual Behavior
This triggers 2 changes:
```hcl
Terraform will perform the following actions:

  ~ module.jumpnodes.azurerm_managed_disk.data_disk
      disk_size_gb:                     "512" => "1024"

  ~ module.jumpnodes.azurerm_virtual_machine.vm
      storage_data_disk.0.disk_size_gb: "512" => "1024"
```
When running apply the disk is succesfully resized and then terraform fails with:

* azurerm_virtual_machine.vm: compute.VirtualMachinesClient#CreateOrUpdate: Failure responding to request: StatusCode=400 -- Original Error: autorest/azure: Service returned an error. Status=400 Code="ResizeDiskError" Message="Managed disk resize via Virtual Machine 'machine' is not allowed. Please resize disk resource at /subscriptions/.../.../.../providers/Microsoft.Compute/disks/data1."

### Steps to Reproduce
Create a machine using an attached managed disk. Increase the size of the attached managed disk and run terraform apply.

**Comment by giskou** [comment]
Behavior seems to be a bit different with Terraform v0.11.0 and provider.azurerm v0.3.3, while still getting an error.

The plan is the same (modify disk and vm) but the error is different:
* azurerm_managed_disk.test: disk.DisksClient#CreateOrUpdate: Failure responding to request: StatusCode=409 -- Original Error: autorest/azure: Service returned an error. Status=409 Code="OperationNotAllowed" Message="Cannot resize disk datadisk_existing while it is attached to running VM /subscriptions/.../providers/Microsoft.Compute/virtualMachines/...."

The way azure suggests those changes should be done is by shutting down the VM, resize the disk and start up again.
In the context of terraform that could be a recreation of the VM but first remove it, then update the disk and finally create the VM again.

Is there a way for that last flow to be done as a workaround?

**Comment by katbyte** [comment]
Hi @jake-delorme, @giskou 

While it would be great to work around this api limitation by shutting down and resizing the disk ourselves it is not high on our priorities. If anyone wants to submit a PR we would be happy to review 🙂

**Comment by r7vme** [comment]
I have similar issue but with `storage_account_type`. I'm switching existing managed disks (attached to VMs) from from `Premium_LRS` to `Standard_LRS`

and it fails with error like that:
```
* azurerm_managed_disk.worker_docker.0: compute.DisksClient#CreateOrUpdate: Failure sending request: StatusCode=409 -- Original Error: failed request: autorest/azure: Service returned an error. Status=<nil> Code="OperationNotAllowed" Message="Cannot change account type, fault domain or network spine of disk gollum-worker-docker-disk-0 while it is attached to running VM /subscriptions/1be3b2e6-497b-45b9-915f-eb35cae23c6a/resourceGroups/gollum/providers/Microsoft.Compute/virtualMachines/worker0."
```

Terraform manifests [here](https://github.com/giantswarm/giantnetes-terraform/blob/master/modules/azure/worker-as/worker-as.tf).

### What i expect?

1. disk detached from VM / or VM recreation forced
2. disk class changed
3. disk attached to VM / or VM created

**Comment by rahulkp220** [comment]
Any updates on this one?

**Comment by jackofallops** [comment]
We recently released v1.43 of the provider which includes an opt-in Beta for the new Virtual Machine and Virtual Machine Scale Set resources coming in 2.0 which supersede the `azurerm_virtual_machine` resource and instead handle attaching Data Disks via the  `azurerm_virtual_machine_managed_disk_attachment` resource.  Changes to data disk size using this mechanism support the necessary VM operations (stop/deallocate etc) to correctly grow the disk when it is attached to a VM.  Since the existing Virtual Machine and Virtual Machine Scale Set resources are being superseded unfortunately we have no plans to backport this fix to the older resources - and as such I’m going to close this issue, however we’d encourage you to check out the new resources.  

Please see https://www.terraform.io/docs/providers/azurerm/guides/2.0-beta.html for beta opt-in details.  

Thanks
-Ste
